File: TP1-parte2/02_Python/count_characters.py
import threading
import time
import sys

def count_characters(lines, start, end, partial_results, index):
    """Counts the number of characters in a subset of lines"""
    count = 0
    for i in range(start, end):
        count += len(lines[i].rstrip('\r\n'))  # quitar saltos de línea
    partial_results[index] = count


def main(file_path, n_threads):
    # Read file into memory
    with open(file_path, "r") as f:
        lines = f.readlines()

    total_lines = len(lines)
    if n_threads > total_lines:
        n_threads = total_lines  # no more threads than lines

    partial_results = [0] * n_threads
    threads = []

    # Start measuring time
    start_time = time.time()

    # Divide work among threads
    block = total_lines // n_threads
    remainder = total_lines % n_threads

    start_line = 0
    for i in range(n_threads):
        end_line = start_line + block + (1 if i < remainder else 0)
        t = threading.Thread(target=count_characters,
                             args=(lines, start_line, end_line, partial_results, i))
        threads.append(t)
        t.start()
        start_line = end_line

    # Wait for all threads
    for t in threads:
        t.join()

    # Sum results
    total_result = sum(partial_results)

    # End of timing and compute durations in milliseconds
    end_time = time.time()
    duration_sec = end_time - start_time           # duration total in seconds
    total_milliseconds = duration_sec * 1000      # convert to milliseconds

    average_one_line_ms = total_milliseconds / total_lines           # average per line
    average_half_lines_ms = total_milliseconds / (total_lines / 2.0) # average per L/2 lines

    # Print results
    print(f"Total Result (number of characters): {total_result}")
    print(f"Total processing time: {total_milliseconds:.3f} ms")
    print(f"Average time per line: {average_one_line_ms:.3f} ms")
    print(f"Average time per L/2 lines: {average_half_lines_ms:.3f} ms")

if __name__ == "__main__":
    file_path = sys.argv[1]
    n_threads = int(sys.argv[2])
    main(file_path, n_threads)

File: TP1-parte2/01_C++/count_characters.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <thread>
#include <chrono>
#include <string>

using namespace std;

// Function that counts characters in a range of lines
void countCharacters(const vector<string>& lines,
                     size_t start, size_t end,
                     long long &partialResult) {
    long long count = 0;
    for (size_t i = start; i < end; ++i) {
        count += lines[i].size();
    }
    partialResult = count;
}

int main(int argc, char* argv[]) {
    string filePath = argv[1];
    int N = stoi(argv[2]);

    // Read the file into memory
    ifstream file(filePath);
    if (!file.is_open()) {
        cerr << "Error: could not open the file.\n";
        return 1;
    }

    vector<string> lines;
    string line;
    while (getline(file, line)) {
        lines.push_back(line);
    }
    file.close();

    size_t totalLines = lines.size();
    if (N > totalLines) N = totalLines; // no more threads than lines

    vector<thread> threads;
    vector<long long> partialResults(N, 0);

    // Start measuring time
    auto startTime = chrono::high_resolution_clock::now();

    // Divide work among threads
    size_t block = totalLines / N;
    size_t remainder = totalLines % N;

    size_t startLine = 0;
    for (int i = 0; i < N; i++) {
        size_t endLine = startLine + block + (i < remainder ? 1 : 0);
        thread t(countCharacters, cref(lines), startLine, endLine, ref(partialResults[i]));
        threads.push_back(move(t));
        startLine = endLine;
    }

    // Wait for all threads to finish
    for (auto &t : threads) {
        t.join();
    }

    // Sum partial results
    long long totalResult = 0;
    for (long long pr : partialResults) {
        totalResult += pr;
    }

    auto endTime = chrono::high_resolution_clock::now();
chrono::duration<double, std::milli> duration_ms = endTime - startTime; // duración en ms
double totalMilliseconds = duration_ms.count();

double average_one_line_ms = totalMilliseconds / totalLines;        // promedio por línea en ms
double average_half_lines_ms = totalMilliseconds / (totalLines / 2.0); // promedio por L/2 líneas en ms

cout << "Total Result (number of characters): " << totalResult << "\n";
cout << "Total processing time: " << totalMilliseconds << " ms\n";
cout << "Average time per line: " << average_one_line_ms << " ms\n";
cout << "Average time per L/2 lines: " << average_half_lines_ms << " ms\n";




    return 0;
}
